monai/data/dataset.py

```diff
@@ -520,13 +520,15 @@ def _load_cache_item(self, idx: int):
             idx: the index of the input data sequence.
         """
         item = self.data[idx]
-        if not isinstance(self.transform, Compose):
+        # copy to make sure thread-safe
+        transform_ = deepcopy(self.transform)
```

Comment:
@rijobro I think this would change the object ids, therefore not compatible with this: https://github.com/Project-MONAI/MONAI/blob/cd08c001a4201ad5d1f03f488cc15938257fe171/monai/transforms/inverse.py#L93 shall we remove the id check?

Reply:
removed in the latest commit, thanks.

Reply:
If we make sure to not modify `self` in `__call__`, then isn't this `deepcopy` superfluous? Having the check against `id(self)` isn't necessary, but I think it's nice if it costs nothing to leave it in.

Reply:
deepcopy is still useful if the dataset is used with some customised transforms. 

the current cachedataset is error-prone, for example https://github.com/Project-MONAI/MONAI/pull/1030, it was quite difficult to locate the issue, and my previous PR was another example of error... https://github.com/Project-MONAI/MONAI/pull/1839#discussion_r600337983

Reply:
Mark: after some online discussion, we aligned to use `deepcopy` in next version, will change back in this PR.
Thanks.